### DISCUSSION TITLE
problem with predicates and tree-sitter-cpp

### DISCUSSION THREAD

**Discussion by BaseMale** [tree-sitter/tree-sitter#3471] [Q&A] [answered]
I'm having trouble getting predicates to work.  In the code below, I'm attempting to write a query that matches the "second_identifier".

I expect the output to look like:
`Match: second_identifier`

But what I get is:
`Match: first_identifier`
`Match: second_identifier`

The same query works in the playground though.  Any ideas what I'm doing wrong?

```cpp
#include <cassert>
#include <iostream>

extern "C" TSLanguage *tree_sitter_cpp(void);

const std::string source = "\
	int first_identifier = 0;\
	int second_identifier = 0;";

const std::string query_string =
    "((identifier) @name (#eq? @name \"second_identifier\"))";

int main() {
  TSLanguage *language = tree_sitter_cpp();
  TSParser *parser = ts_parser_new();
  ts_parser_set_language(parser, language);

  uint32_t error_offset;
  TSQueryError error_type;
  TSQuery *query =
      ts_query_new(language, query_string.c_str(), query_string.length(),
                   &error_offset, &error_type);

  TSTree *tree =
      ts_parser_parse_string(parser, nullptr, source.c_str(), source.length());

  TSNode root = ts_tree_root_node(tree);
  
  TSQueryCursor *query_cursor = ts_query_cursor_new();
  ts_query_cursor_exec(query_cursor, query, root);

  TSQueryMatch match;
  while (ts_query_cursor_next_match(query_cursor, &match)) {
    assert(match.capture_count == 1);

    TSNode match_node = match.captures[0].node;
    const uint32_t start_byte = ts_node_start_byte(match_node);
    const uint32_t end_byte = ts_node_end_byte(match_node);
    const std::string name = source.substr(start_byte, end_byte - start_byte);
    
    std::cout << "Match: " << name << std::endl;
  }

  ts_query_cursor_delete(query_cursor);
  ts_tree_delete(tree);
  ts_query_delete(query);
  ts_parser_delete(parser);

  return 0;
}
```

**Comment by BaseMale** [comment, marked as answer]
After reading through the code a bit, it seems like predicates aren't handled within tree-sitter's query interface.  The rust and javascript bindings handle them post query op.  To do this from within a c/cpp program like above, you'd handle them yourself as you iterate through the matches

**Reply by sogaiu** [reply]
Possibly of interest:

> Note — Predicates are not handled directly by the Tree-sitter C library. They are just exposed in a structured form so that higher-level code can perform the filtering. 

(text from near the end of the [Query Syntax section](https://tree-sitter.github.io/tree-sitter/using-parsers#query-syntax) of the website docs)

**Reply by BaseMale** [reply]
If only I could read :man_facepalming:

**Reply by sogaiu** [reply]
Obligatory [Hitchhiker's Guide reference](https://www.azquotes.com/quote/354892) (^^;
